lib/src/markdown_processor.dart

```diff
@@ -207,102 +267,369 @@ MatchingLinkResult _getMatchingLinkElement(
+      }
+    }
+
+    for (Class tryClass in tryClasses) {
+      if (tryClass != null) {
+        _getResultsForClass(
+            tryClass, codeRefChomped, results, codeRef, package);
+      }
+      if (results.isNotEmpty) break;
+    }
+    // Sometimes documentation refers to classes that are further up the chain.
+    // Get those too.
+    if (results.isEmpty && realClass != null) {
+      for (Class superClass
+          in realClass.superChain.map((et) => et.element as Class)) {
+        if (!tryClasses.contains(superClass)) {
+          _getResultsForClass(
+              superClass, codeRefChomped, results, codeRef, package);
+        }
+        if (results.isNotEmpty) break;
+      }
+    }
+  }
+  results.remove(null);
+
+  // We now need the ref element cache to keep from repeatedly searching [Package.allModelElements].
+  // TODO(jcollins-g): Find somewhere to cache elements outside package.libraries
+  //                   so we can give the right warning (no canonical found)
+  //                   when referring to objects in libraries outside the
+  //                   documented set.
+  if (results.isEmpty && _findRefElementCache == null) {
+    assert(package.allLibrariesAdded);
+    _findRefElementCache = new Map();
+    for (final modelElement in package.allModelElements) {
+      _findRefElementCache.putIfAbsent(
+          modelElement.fullyQualifiedNameWithoutLibrary, () => new Set());
+      _findRefElementCache.putIfAbsent(
+          modelElement.fullyQualifiedName, () => new Set());
+      _findRefElementCache[modelElement.fullyQualifiedName].add(modelElement);
+      _findRefElementCache[modelElement.fullyQualifiedNameWithoutLibrary]
+          .add(modelElement);
+    }
+  }
+
+  // But if not, look for a fully qualified match.  (That only makes sense
+  // if the codeRef might be qualified, and contains periods.)
+  if (results.isEmpty &&
+      codeRefChomped.contains('.') &&
+      _findRefElementCache.containsKey(codeRefChomped)) {
+    for (final modelElement in _findRefElementCache[codeRefChomped]) {
+      if (!_ConsiderIfConstructor(codeRef, modelElement)) continue;
+      results.add(package.findCanonicalModelElementFor(modelElement.element));
     }
   }
+  results.remove(null);
 
   // Only look for partially qualified matches if we didn't find a fully qualified one.
-  if (result.isEmpty) {
-    for (final modelElement in library.allCanonicalModelElements) {
-      if (codeRef == modelElement.fullyQualifiedNameWithoutLibrary &&
-          modelElement is! Constructor) {
-        result[modelElement.fullyQualifiedName] = modelElement;
+  if (results.isEmpty) {
+    for (final modelElement in library.allModelElements) {
+      if (!_ConsiderIfConstructor(codeRef, modelElement)) continue;
+      if (codeRefChomped == modelElement.fullyQualifiedNameWithoutLibrary) {
+        results.add(package.findCanonicalModelElementFor(modelElement.element));
+      }
+    }
+  }
+  results.remove(null);
```

Comment:
I just want to make sure these `results.remove(null);` are intentional, and something we should comment about, or something we should eventually fix.